command/013_config_upgrade.go

```diff
@@ -138,64 +470,79 @@ necessary adjustments, and then commit.
 	return 0
 }
 
-// For providers which need a source attribute, detect and return source
-// FIXME: currently does not filter or detect sources
-func detectProviderSources(providers moduledeps.Providers) (map[string]string, tfdiags.Diagnostics) {
-	sources := make(map[string]string)
-	for provider := range providers {
-		sources[provider.Type] = provider.String()
-	}
-	return sources, nil
-}
-
-var providersTemplate = template.Must(template.New("providers.tf").Parse(`terraform {
-  required_providers {
-    {{- range $type, $source := .}}
-    {{$type}} = {
-      source = "{{$source}}"
-    }
-    {{- end}}
-  }
-}
-`))
-
-// Generate a file with terraform.required_providers blocks for each provider
-func generateRequiredProviders(providerSources map[string]string, dir string) tfdiags.Diagnostics {
+// For providers which need a source attribute, detect the source
+func (c *ZeroThirteenUpgradeCommand) detectProviderSources(requiredProviders map[string]*configs.RequiredProvider) tfdiags.Diagnostics {
+	source := c.providerInstallSource()
 	var diags tfdiags.Diagnostics
 
-	// Find unused file named "providers.tf", or fall back to e.g. "providers-1.tf"
-	path := filepath.Join(dir, "providers.tf")
-	if _, err := os.Stat(path); !os.IsNotExist(err) {
-		for i := 1; ; i++ {
-			path = filepath.Join(dir, fmt.Sprintf("providers-%d.tf", i))
-			if _, err := os.Stat(path); os.IsNotExist(err) {
-				break
+	for name, rp := range requiredProviders {
+		// If there's already an explicit source, skip it
+		if rp.Source != "" {
+			continue
+		}
+
+		// Construct a legacy provider FQN using the existing addr's type. This
+		// is necessary because the config parser for required providers
+		// constructs a default provider FQN for configurations with no source.
+		// For this tool specifically we want to treat those as legacy
+		// providers, so that we can look up the namespace on the registry.
+		addr := addrs.NewLegacyProvider(rp.Type.Type)
```

Comment:
If I'm remembering right (which I might not be) I think `name` here is what we're now calling the "local name" but that was formerly the "type name", which would be "aws" in this example:

```hcl
  required_providers {
    aws = "~> 2.0.0"
  }
```

Would it make sense to use `name` to construct the legacy provider here? (Maybe this is moot because `rp.Type.Type` was already populated from that name anyway? If so, don't worry about this.)

Reply:
`rp.Type.Type` _was_ already populated from the local name. But in the case of `required_providers` blocks, that happens quite some distance away from this file. I think being explicit here about generating the legacy FQN from the local name is clearer, so I've made that change.